File: 2021/09/14/DistributeCandies.java
/**
 * https://leetcode.com/problems/distribute-candies/
 * @author Andrew Jarombek
 * @since 9/14/2021
 */

import java.util.HashSet;
import java.util.Set;

public class DistributeCandies {
    public int distributeCandies(int[] candyType) {
        Set<Integer> candyTypes = new HashSet<>();
        int uniqueCandies = 0;

        for (int candy: candyType) {
            if (!candyTypes.contains(candy)) {
                uniqueCandies++;
                candyTypes.add(candy);
            }
        }

        return Math.min(uniqueCandies, candyType.length / 2);
    }
}

File: 2022/09/14/distributeCandies.js
/**
 * https://leetcode.com/problems/distribute-candies/
 * @author Andrew Jarombek
 * @since 9/14/2022
 */

const distributeCandies = function(candyType) {
    const set = new Set();

    for (let candy of candyType)
        set.add(candy);

    return Math.min(set.size, Math.floor(candyType.length / 2));
};

console.assert(distributeCandies([1,1,2,2,3,3]) === 3);
console.assert(distributeCandies([1,1,2,3]) === 2);
console.assert(distributeCandies([1,1,1,1]) === 1);